cpp/src/parquet/arrow/path_internal_test.cc

```diff
@@ -0,0 +1,426 @@
+  auto list_field = field("list", list(entries), /*nullable=*/true);
+  auto nested_list_type = list(list_field);
+  auto double_nested_list_type = list(nested_list_type);
+
+  auto array = ::arrow::ArrayFromJSON(double_nested_list_type,
+                                      R"([
+                                           [null, [[1 , null, 3], []], []], 
+                                           [[[]], [[], [1, 2]], null, [[3]]],
+					   null, 
+ 					   []
+                                         ])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>{2, 7, 6, 7, 5, 3,  // first row
+                                                         5, 5, 7, 7, 2, 7,  // second row
+                                                         0,                 // third row
+                                                         1},
+                     /*rep_levels=*/std::vector<int16_t>{0, 1, 3, 3, 2, 1,  // first row
+                                                         0, 1, 2, 3, 1, 1,  // second row
+                                                         0, 0});
+}
+
+TEST_F(MultipathLevelBuilderTest, TestStruct) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_field = field("list", list(entries), /*nullable=*/true);
+  auto struct_type = ::arrow::struct_({list_field, entries});
+
+  auto array = ::arrow::ArrayFromJSON(struct_type,
+                                      R"([{"Entries" : 1, "list": [2, 3]}, 
+                                          {"Entries" : 4, "list": [5, 6]},
+                                          null])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+  ASSERT_THAT(results_, SizeIs(2));
+  results_[0].CheckLevels(/*def_levels=*/std::vector<int16_t>{4, 4, 4, 4, 0},
+                          /*rep_levels=*/std::vector<int16_t>{0, 1, 0, 1, 0});
+  results_[1].CheckLevelsWithNullRepLevels(
+      /*def_levels=*/std::vector<int16_t>({2, 2, 0}));
+}
+
+TEST_F(MultipathLevelBuilderTest, TestFixedSizeList) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/false);
+  auto list_type = fixed_size_list(entries, 2);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([null, [2, 3], [4, 5], null])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  results_[0].CheckLevels(/*def_levels=*/std::vector<int16_t>{0, 1, 1, 1, 1, 0},
+                          /*rep_levels=*/std::vector<int16_t>{0, 0, 1, 0, 1, 0});
+
+  ASSERT_THAT(results_[0].post_list_elements, SizeIs(1));
+  EXPECT_THAT(results_[0].post_list_elements[0].start, Eq(2));
+  EXPECT_THAT(results_[0].post_list_elements[0].end, Eq(6));
```

Comment:
I am understanding what this member is better now, but for the slow pokes (i.e. such as me) may be worth saying that the null slots take up space in a fixed size list and so the actual "written" values are the middle 4 elements